src/super_gradients/training/sg_model/sg_model.py

```diff
@@ -1653,3 +1669,41 @@ def evaluate(self, data_loader: torch.utils.data.DataLoader, metrics: MetricColl
         if self.multi_gpu == MultiGPUMode.DISTRIBUTED_DATA_PARALLEL:
             logging_values = reduce_results_tuple_for_ddp(logging_values, next(self.net.parameters()).device)
         return logging_values
+
+    @staticmethod
+    def instantiate_net(architecture: Union[torch.nn.Module, SgModule.__class__, str], arch_params: dict,
+                        checkpoint_params: dict, *args, **kwargs) -> tuple:
+        """
+        Instantiates nn.Module according to architecture and arch_params, and handles pretrained weights and the required
+            module manipulation (i.e head replacement).
+
+        :param architecture: String, torch.nn.Module or uninstantiated SgModule class describing the netowrks architecture.
+        :param arch_params: Architecture's parameters passed to networks c'tor.
+        :param checkpoint_params: checkpoint loading related parameters dictionary with 'pretrained_weights' key,
+            s.t it's value is a string describing the dataset of the pretrained weights (for example "imagenent").
+
+        :return: instantiated netowrk i.e torch.nn.Module, architecture_class (will be none when architecture is not str)
+
+        """
+        pretrained_weights = core_utils.get_param(checkpoint_params, 'pretrained_weights', default_val=None)
+
+        if pretrained_weights is not None:
+            num_classes_new_head = arch_params.num_classes
+            arch_params.num_classes = PRETRAINED_NUM_CLASSES[pretrained_weights]
+
+        architecture_cls = None
+        if isinstance(architecture, str):
+            architecture_cls = ARCHITECTURES[architecture]
+            net = architecture_cls(arch_params=arch_params)
+        elif isinstance(architecture, SgModule.__class__):
+            net = architecture(arch_params)
+        else:
+            net = architecture
+
+        if pretrained_weights:
+            load_pretrained_weights(net, architecture, pretrained_weights)
+            if num_classes_new_head != arch_params.num_classes:
+                net.replace_head(new_num_classes=num_classes_new_head)
+                arch_params.num_classes = num_classes_new_head
+
+        return net, architecture_cls
```

Comment:
@shaydeci  - Why are we returning both net and architecture_cls?

Running the parsing of the str:
architecture_cls = ARCHITECTURES[architecture] is not relevant any more...

The only usage of self.architecture_cls is in re_build_model... (line 1234) - but we can (and probably should) replace this call with instantiate_net...

Reply:
As discussed- we need to return them as the nets are used for initialising the teacher and student (KDModel.instantiate_net())

Reply:
@shaydeci  - Why do we need to return architecture_cls? We need to change re_build_model to use instantiate_net... And we can remove the self.architecture_cls model class attribute since it is not needed anymore...

Reply:
done